sys/net/gnrc/application_layer/tftp/gnrc_tftp.c

```diff
@@ -0,0 +1,848 @@
+
+/* send data or and ack depending if we are reading or writing */
+static tftp_state _tftp_send_dack(tftp_context_t *ctxt, gnrc_pktsnip_t *buf, tftp_opcodes_t op);
+
+/* send and TFTP error to the client */
+static tftp_state _tftp_send_error(tftp_context_t *ctxt, gnrc_pktsnip_t *buf, tftp_err_codes_t err, const char *err_msg);
+
+/* this function sends the actual packet */
+static tftp_state _tftp_send(gnrc_pktsnip_t *buf, tftp_context_t *ctxt, size_t len);
+
+/* decode the default TFTP start packet */
+static int _tftp_decode_start(tftp_context_t *ctxt, uint8_t *buf, gnrc_pktsnip_t *outbuf);
+
+/* decode the TFTP option extensions */
+static  int _tftp_decode_options(tftp_context_t *ctxt, gnrc_pktsnip_t *buf, uint32_t start);
+
+/* decode the received ACK packet */
+static bool _tftp_validate_ack(tftp_context_t *ctxt, uint8_t *buf);
+
+/* processes the received data packet and calls the callback defined by the user */
+static int _tftp_process_data(tftp_context_t *ctxt, gnrc_pktsnip_t *buf);
+
+/* decode the received error packet and calls the callback defined by the user */
+static int _tftp_decode_error(uint8_t *buf, tftp_err_codes_t *err, const char **err_msg);
+
+/* get the maximum allowed transfer unit to avoid 6Lo fragmentation */
+static uint16_t _tftp_get_maximum_block_size(void) {
+    uint16_t tmp;
+    kernel_pid_t ifs[GNRC_NETIF_NUMOF];
+    size_t ifnum = gnrc_netif_get(ifs);
+
+    return 10;
+
+    if (ifnum > 0 && gnrc_netapi_get(ifs[0], NETOPT_MAX_PACKET_SIZE, 0, &tmp, sizeof(uint16_t)) >= 0) {
+        /* TODO calculate proper block size */
+        return tmp - sizeof(udp_hdr_t) - sizeof(ipv6_hdr_t) - 50;
+    }
+
+    return GNRC_TFTP_MAX_TRANSFER_UNIT;
+}
+
+int gnrc_tftp_client_read(ipv6_addr_t *addr, const char *file_name,
+                          tftp_data_callback data_cb, tftp_transfer_start_callback start_cb) {
+    tftp_context_t ctxt;
+
+    /* prepare the context */
+    if (_tftp_init_ctxt(addr, file_name, start_cb, data_cb, TO_RRQ, &ctxt) != FINISHED) {
+        return -EINVAL;
+    }
+
+    /* set the transfer options */
+    uint16_t mtu = _tftp_get_maximum_block_size();
+    if (_tftp_set_opts(&ctxt, mtu, 1, 0) != FINISHED) {
+        return -EINVAL;
+    }
+
+    /* start the process */
+    return _tftp_do_client_transfer(&ctxt);
+}
+
+int gnrc_tftp_client_write(ipv6_addr_t *addr, const char *file_name,
+                           tftp_data_callback data_cb, uint32_t total_size) {
+    tftp_context_t ctxt;
+
+    /* prepare the context */
+    if (_tftp_init_ctxt(addr, file_name, NULL, data_cb, TO_RWQ, &ctxt) < 0) {
+        return -EINVAL;
+    }
+
+    /* set the transfer options */
+    uint16_t mtu = _tftp_get_maximum_block_size();
+    if (_tftp_set_opts(&ctxt, mtu, 1, total_size) != FINISHED) {
+        return -EINVAL;
+    }
+
+    /* start the process */
+    return _tftp_do_client_transfer(&ctxt);
+}
+
+#define INT_DIGITS 6
+char *itoa(int i, size_t *len) {
```

Comment:
This seems to be copy-pasted from somewhere on the internet. There is [this](http://opensource.apple.com/source/groff/groff-10/groff/libgroff/itoa.c) for example which is released under GPL. AFAICT this is not a good idea because strictly speaking it would render all RIOT code also GPL.

On the other hand, why not using `itoa` from standard library?